Feature/eckit fraction safe casts

pmaciel opened the pull request:
This is a port of Atlas's FPE trapping, quite modified in other directions (no parallel logging, minimally intrusive signal handling). The objective is to trap FPEs earlier in the development, particularly eckit::Fraction, which is being a big source of runtime issues for user with unusual cases.

I've included Kay's suggestion of an appropriate "safe casting" missing in the recent eckit::Fraction. There's urgency on this because of the impeding sync and earthkit-* releases, akthough there are higher priority issues at the moment.

I suggest to review this PR as all the final changes at once (ie. the individual commits might be too much to look at.)

wdeconinck commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -0,0 +1,337 @@
+
+
+class Signals : std::map<int, Signal> {
+    // Not sure if this should be made public (in header file) just yet
+
+    static Signals& instance() {
+        static Signals signals;
+        return signals;
+    }
+
+    void setSignalHandler(int signum) {
+        if (find(signum) != end()) {
+            restoreSignalHandler(signum);
+        }
+
+        const auto& signal = (operator[](signum) = Signal{signum});
+        sigaction(signum, signal.action(), nullptr);
+    }
+
+    void restoreSignalHandler(int signum) {
+        if (auto it = find(signum); it != end()) {
+            std::signal(signum, SIG_DFL);
+            erase(signum);
+        }
+    }
+
+    void restoreAllSignalHandlers() {
+        while (!empty()) {
+            restoreSignalHandler(rbegin()->first);
+        }
+    }
+
+public:
+
+    static void set_signal_handler(int signum) { instance().setSignalHandler(signum); }
+    static void restore_signal_handler(int signum) { instance().restoreSignalHandler(signum); }
+    static void restore_all_signal_handlers() { instance().restoreAllSignalHandlers(); }
+
+    static const Signal& signal(int signum) { return instance().at(signum); }
+};
```
This code is ported from Atlas, where it is more used, controllable via environment variables. Hence the generic code with map of string to int. Very likely not everything is needed, but it could serve as a basis to be used more generally later.

wdeconinck commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -0,0 +1,337 @@
+    Log::error() << "[Signal: " << signal << ", Floating Point Exception: " << code << "]" << std::endl;
+
+    Signals::restore_all_signal_handlers();
+    LibEcKit::instance().abort();
```
The reason to use eckit's abort is because it may call MPI_Abort, which we need in a MPI-parallel context. Normally it calls std::abort instead.

Ozaq replied:
I don't have MPI experience beyond basic examples but its that detection of process crashes and shutdown one oft the responsibilities of the MPI process manager? I.e. is MPI_ABORT actually necessary to prevent hangs because the MPI-PM will end all remaining processes gracefully on failure? Or at least could be configured to do so.

wdeconinck replied:
When you don't call MPI_Abort, then other processes are just continuing happily, and upon an MPI receive or barrier it may just wait forever and hang or wait for a timeout.
The MPI_Abort essentially triggers some kind of SIGABRT on the other MPI processes, and lets those know from which MPI process MPI_Abort was called.
Then teardown happens everywhere quasi-simultaneously without hangs.

Ozaq replied:
@wdeconinck yes I understand that, but I was under the impression that this issue can also be solved by the watchdog capability of the MPI process manager that checks if all processes are still up and terminates all if a process failed.

I.e. My understanding is that if we use HYDRA, the default behavior is to kill all processes anyways if one fails. So why bother with the MPI_ABORT in the first place. I assume the answer is because we do not use HYDRA or it is configured differently. Right?

Just to be clear: I am not blocking the PR because of the call to `MPI_ABORT` just trying to understand the details here.

wdeconinck commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -0,0 +1,337 @@
+    volatile double x = 0.;
+    volatile double y = 0.;
+    volatile double z = x / y;
+    (void)z;  // prevent optimization
+}
+
+
+void FloatingPointExceptions::test_inexact() {
+    volatile double x = 1 / 3.;
+    volatile double y = x * 3.;
+    (void)y;
+}
+
+
+void FloatingPointExceptions::test_divbyzero() {
+    volatile double x = 1.;
+    volatile double y = 0.;
+    volatile double z = x / y;
+    (void)z;
+}
+
+
+void FloatingPointExceptions::test_overflow() {
+    volatile double x = 1.e308;
+    volatile double y = x * x;
+    (void)y;
+}
+
+
+void FloatingPointExceptions::test_underflow() {
+    volatile double x = 1.e-308;
+    volatile double y = x * x;
+    (void)y;
+}
+
+
+void FloatingPointExceptions::test_denormal() {
+    volatile double x = 1.e-320;
+    (void)x;
+}
```
This could reside in unit-test indeed.

Ozaq replied:
Alright, in that case please just this short explanation as a comment for the next guy who wonders why they are not in the tests :)

pmaciel commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -0,0 +1,337 @@
+
+int make_excepts(const std::string& codes) {
+    static const std::map<std::string, int> CODE_TO_EXCEPT{
+        {"FE_INVALID", FE_INVALID},   {"FE_INEXACT", FE_INEXACT},     {"FE_DIVBYZERO", FE_DIVBYZERO},
+        {"FE_OVERFLOW", FE_OVERFLOW}, {"FE_UNDERFLOW", FE_UNDERFLOW}, {"FE_ALL_EXCEPT", FE_ALL_EXCEPT}};
```
Yes indeed, and I did consider it, in what would essentialy be a platform-independent way to describe FPEs. I didn't see the bitflag opportunity but that looks nice, @wdeconinck already has done some work like this. 

The number of useful FPEs is however rather small and the handling of 'excepts' (where bitmasks would finally see the most use) is deliberately very low level (C-style bit masking). const char*/std::string is the absolute minimum for writing information to the user -- so I stuck to just that minimum :-) (a future development, perhaps).

Ozaq replied:
Note to self: Create eckit issue

wdeconinck commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -0,0 +1,337 @@
+    }
+#endif
+
+    Log::error() << "[Signal: " << signal << ", Floating Point Exception: " << code << "]" << std::endl;
```
You mean https://cplusplus.com/reference/ostream/ostream/write/ ?

wdeconinck commented on src/eckit/maths/FloatingPointExceptions.h:
```diff
@@ -0,0 +1,32 @@
+class FloatingPointExceptions {
+public:
+
+    static void enable_floating_point_exceptions(const std::string& = "FE_DIVBYZERO, FE_INVALID, FE_OVERFLOW");
+    static void disable_floating_point_exceptions(const std::string& = "FE_ALL_EXCEPT");
```
To stress once more, there is code in Atlas, and ectrans and other places that sometimes wants to enable/disable FPE in the context of a routine when too aggressive optimisation with branch prediction and vectorisation causes e.g. a FE_DIVBYZERO or FE_INVALID, which later gets masked out, but the SIGFPE gets thrown regardless. The only other solution to this would be to reduce optimisation level, or use volatile keywords etc preventing optimisation, which we want to avoid.
It would be undesirable to start messing with the signal handler here.

pmaciel commented:
I think I've addressed all concerns raised here -- can you re-check?

The changes are

1. strictly no allocation on signal handling,
2. flattening of signal handling hierarchy/structure
3. possible report of multiple FPEs
4. No attempt made for compatible messages (on a Apple Silicon, SIGILL ESR=DZF does not report "FE_DIVBYZERO", just "SIGILL" + "DZF".)

Ozaq commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -0,0 +1,330 @@
+[[noreturn]] void custom_signal_handler(int signum, ::siginfo_t* si, [[maybe_unused]] void* ucontext);
+
+
+// TODO use/improve eckit::SignalHandler
```
Note to self: Create eckit issue

Ozaq commented on src/eckit/maths/FloatingPointExceptions.h:
```diff
@@ -0,0 +1,65 @@
+ * @class FloatingPointExceptions
+ * @brief Manage floating-point exception handling via signal handlers.
+ *
+ * Enabling floating-point exceptions causes the program to raise signals (e.g., SIGFPE) on events such as division by
+ * zero or overflow, allowing detection and debugging of numerical issues.
```
It misses one crucial point: The FPE is handled by aborting the process.

pmaciel commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -0,0 +1,330 @@
+        //    https://developer.arm.com/documentation/ddi0595/2020-12/AArch64-Registers/ESR-EL1--Exception-Syndrome-Register--EL1-
+        //    https://developer.arm.com/documentation/ddi0595/2020-12/AArch64-Registers/ESR-EL2--Exception-Syndrome-Register--EL2-
+
+        const std::bitset<32> esr = reinterpret_cast<ucontext_t*>(ucontext)->uc_mcontext->__es.__esr;
+        constexpr std::bitset<32> fpe_mask(0b10110000000000000000000000000000);
```
Yes

wdeconinck commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -0,0 +1,353 @@
+
+
+int FloatingPointExceptions::excepts() {
+    return CUSTOM_EXCEPTS;
```
This would only return the FE codes that were set using these functions, but not the ones that could have been set by other calls to feenableexcept, or compiler flags, or Fortran intrinsics.

pmaciel replied:
Yes and neither does it respect the exceptions set via other means

wdeconinck reviewed (commented):
Thank you @pmaciel  for addressing the review comments.

The new code about the RestorableEnvironment is a bit magic, and I can't claim to understand it. Especially I'm not sure why it is required to raise excepts and sigmp.
The local variable `int ex=0;` in `RestorableEnvironment::excepts`, seems initialized to zero but then not touched, and returned after some magic happens.

Is there any current platform we are running the CI on that covers this code path?

pmaciel commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -247,32 +264,114 @@ void disable_floating_point_exception(int excepts) {
+
+        ::sigaction(SIGFPE, &saved_act, nullptr);
+
+        return ex;
```
There was a mstake there, at the beggining of the function "int ex = 0;", it should just be "ex = 0" -- leftover from debugging.

I was just trying to cover your request or separating the enabling/disabling of FPEs from the enabling/disabling of signal handlers. But, it is full of pitfalls and certainly not what I'm trying to do here.

pmaciel commented on src/eckit/maths/FloatingPointExceptions.cc:
```diff
@@ -247,32 +264,114 @@ void disable_floating_point_exception(int excepts) {
+    return ex;
+#else
+    RestorableEnvironment env;
+    return env.excepts();
```
I've removed this in favour of directly supporting Intel x86/x86_64 and Apple x86_64/arm64. My attempt of reimplementing fegetenv was full of pitfalls, and supporting too much flaky functionality.

I've also restored the simpler interface 
void enable_floating_point_exceptions(const std::string& names)
void disable_floating_point_exceptions(const std::string& names)

which tracks which exceptions have been set by this mechanism. The signal handlers also backup any existing signal action and replace it with the custom handler, and on disabling the signal handler gets restored. This should play nicely with other methods including std:: ones.

Currently this is enough to raise/catch FPEs on the common platforms -- I don't want to do more than necessary.

pmaciel commented:
> I don't agree to overwrite user-specified signal handler with any custom one unless specified explicitly with an api call as implemented just before your last revert, and only if the end-user is aware and can control this. Imagine this program:
> 
> ```c++
> int main() {
>    install_my_signal_handler();   // e.g. DR_HOOK
>  
>     // some of eckit code, or whatever code that will be using this PRs API
>     {    
>         eckit::maths::FloatingPointExceptions::enable_floating_point_exceptions( FE_INVALID );
>         // now eckit custom signal handler was installed unexpectedly, and likely undesired.
>         std::sqrt(-1.);
>         eckit::maths::FloatingPointExceptions::disable_floating_point_exceptions( );
>      }
> }
> ```
> 
> Now my installed signal handler was not used, and instead the one from eckit, which is not going to be doing what is required. Perhaps DR_HOOK wants to dump a corefile Restoring to the previous is not enough, because the abort will have happened.

I disagree -- in the previous case that I replace the signal handler, then you'll never have the behaviour back. This was you can isolate FPE-sensitive parts and restore the previous behaviour, including whatere DR_HOOK would do -- inclusing even recovering from whicever signal would possibly be sent. This code is very limited, to SIGFPE only (on macs, also to SIGILL which is more generic).

The other option, as before, is to set to default signal handling, which would destroy whichever behaviour was set before, and also doesn't help.

> fedisableexcept

Thanks -- I hadn't found a platform that had these enabled

pmaciel commented:
What's the situation -- I believe all isues are resolved?

wdeconinck reviewed (approved):
👍 Very nice now! thanks all for the thorough process this went through.

pmaciel commented:
I'm not sure what happened here -- the branch has been merged cleanly into develop ([20b09aef](https://github.com/ecmwf/eckit/commit/20b09aef)), certainly not what the message says ("Closed with unmerged commits") -- that would be a disaster. But, it looks alright!

wdeconinck commented:
> I'm not sure what happened here -- the branch has been merged cleanly into develop ([20b09aef](https://github.com/ecmwf/eckit/commit/20b09aef)), certainly not what the message says ("Closed with unmerged commits") -- that would be a disaster. But, it looks alright!

👌 that just happens when not using Github's buttons :) no problem.
